File: src/main/kotlin/me/hippodev/routing/BackendPinger.kt
package me.hippodev.routing

import io.netty.bootstrap.Bootstrap
import io.netty.buffer.ByteBuf
import io.netty.buffer.Unpooled
import io.netty.channel.ChannelFutureListener
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInitializer
import io.netty.channel.ChannelOption
import io.netty.channel.EventLoopGroup
import io.netty.channel.socket.SocketChannel
import io.netty.channel.socket.nio.NioSocketChannel
import io.netty.handler.codec.ByteToMessageDecoder
import me.hippodev.protocol.encodeHandshake
import me.hippodev.protocol.encodeProxyProtocolHeader
import me.hippodev.protocol.readFrameLength
import me.hippodev.protocol.readString
import me.hippodev.protocol.readVarInt
import me.hippodev.protocol.writeVarInt
import java.net.InetAddress
import java.net.InetSocketAddress
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

data class PingResult(val statusJson: String, val latencyMillis: Long)

/**
 * Opens a fresh connection to [addr] and performs a real Server List Ping
 * (handshake + status request) right now, independent of any client traffic
 * or the [PingCache]. Used for on-demand health checks (e.g. the status API)
 * rather than the passive, client-triggered pings in StatusHandler.
 */
fun pingBackendLive(
    eventLoopGroup: EventLoopGroup,
    addr: InetSocketAddress,
    protocolVersion: Int,
    virtualHost: String,
    port: Int,
    timeoutMillis: Long = 5000,
    proxyProtocol: Boolean = false
): CompletableFuture<PingResult> {
    val future = CompletableFuture<PingResult>()
    val startTime = System.currentTimeMillis()
    var openChannel: io.netty.channel.Channel? = null

    val bootstrap = Bootstrap()
        .group(eventLoopGroup)
        .channel(NioSocketChannel::class.java)
        .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, timeoutMillis.toInt())
        .handler(object : ChannelInitializer<SocketChannel>() {
            override fun initChannel(ch: SocketChannel) {
                ch.pipeline().addLast(object : ByteToMessageDecoder() {
                    override fun decode(ctx: ChannelHandlerContext, buf: ByteBuf, out: MutableList<Any>) {
                        val frameStart = buf.readerIndex()
                        val length = try {
                            readFrameLength(buf, frameStart)
                        } catch (e: IllegalStateException) {
                            future.completeExceptionally(e)
                            ctx.close(); return
                        } ?: return
                        if (buf.readableBytes() < length) {
                            buf.readerIndex(frameStart); return
                        }
                        val packetId = readVarInt(buf)
                        if (packetId == 0x00) {
                            val json = readString(buf, 262144)
                            future.complete(PingResult(json, System.currentTimeMillis() - startTime))
                        } else {
                            future.completeExceptionally(IllegalStateException("Unexpected packet id $packetId from backend"))
                        }
                        ctx.close()
                    }
                })
            }
        })

    bootstrap.connect(addr).addListener(ChannelFutureListener { connectFuture ->
        if (!connectFuture.isSuccess) {
            future.completeExceptionally(connectFuture.cause() ?: IllegalStateException("connect failed"))
            return@ChannelFutureListener
        }
        val channel = connectFuture.channel()
        openChannel = channel
        if (proxyProtocol) {
            // No real client behind this probe (it's a standalone health check) - a placeholder
            // source is fine, backends enforcing proxyProtocol just need a syntactically valid
            // header before they'll process anything past it.
            val source = InetSocketAddress(InetAddress.getLoopbackAddress(), 0)
            channel.writeAndFlush(encodeProxyProtocolHeader(source, addr))
        }
        channel.writeAndFlush(encodeHandshake(protocolVersion, virtualHost, port, 1))
        val requestFrame = Unpooled.buffer()
        writeVarInt(requestFrame, 1)
        writeVarInt(requestFrame, 0x00)
        channel.writeAndFlush(requestFrame)
    })

    eventLoopGroup.next().schedule({
        if (future.completeExceptionally(TimeoutException("Ping to $addr timed out after ${timeoutMillis}ms"))) {
            openChannel?.close()
        }
    }, timeoutMillis, TimeUnit.MILLISECONDS)

    return future
}

File: src/backend_pinger.rs
//! Port of `routing/BackendPinger.kt` — dials a backend right now and performs a real Server
//! List Ping, independent of any client traffic or the `PingCache`. Used for on-demand health
//! checks (the API's live `/ping` endpoint) as well as `server.rs`'s client-triggered status path
//! (which additionally validates and caches the result — see `ping_cache.rs`).

use std::net::SocketAddr;
use std::time::{Duration, Instant};

use tokio::net::TcpStream;

use crate::buffered_stream::BufferedStream;
use crate::minecraft_protocol::{encode_handshake, encode_proxy_protocol_header};
use crate::varint::{read_var_int, read_string, write_var_int};

pub struct PingResult {
    pub status_json: String,
    pub latency_millis: i64,
}

#[derive(Debug)]
pub enum PingError {
    Connect(std::io::Error),
    Timeout,
    Io(std::io::Error),
    UnexpectedPacketId(i32),
}

impl std::fmt::Display for PingError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PingError::Connect(e) => write!(f, "connect failed: {e}"),
            PingError::Timeout => write!(f, "ping timed out"),
            PingError::Io(e) => write!(f, "io error: {e}"),
            PingError::UnexpectedPacketId(id) => write!(f, "unexpected packet id {id} from backend"),
        }
    }
}
impl std::error::Error for PingError {}

/// Opens a fresh connection to `addr` and performs a real Server List Ping (handshake + status
/// request) right now.
pub async fn ping_backend_live(
    addr: SocketAddr,
    protocol_version: i32,
    virtual_host: &str,
    port: u16,
    timeout: Duration,
    proxy_protocol: bool,
) -> Result<PingResult, PingError> {
    let start = Instant::now();
    let connect = tokio::time::timeout(timeout, TcpStream::connect(addr)).await;
    let mut stream = match connect {
        Ok(Ok(s)) => s,
        Ok(Err(e)) => return Err(PingError::Connect(e)),
        Err(_) => return Err(PingError::Timeout),
    };
    let _ = stream.set_nodelay(true);

    if proxy_protocol {
        // No real client behind this probe (it's a standalone health check) - a placeholder
        // loopback source is fine, backends enforcing proxyProtocol just need a syntactically
        // valid header before they'll process anything past it.
        let source: SocketAddr = "127.0.0.1:0".parse().unwrap();
        let header = encode_proxy_protocol_header(source, addr);
        tokio::io::AsyncWriteExt::write_all(&mut stream, &header).await.map_err(PingError::Io)?;
    }
    let hs = encode_handshake(protocol_version, virtual_host, port, 1);
    tokio::io::AsyncWriteExt::write_all(&mut stream, &hs).await.map_err(PingError::Io)?;
    let mut request = Vec::new();
    write_var_int(&mut request, 1);
    write_var_int(&mut request, 0x00);
    tokio::io::AsyncWriteExt::write_all(&mut stream, &request).await.map_err(PingError::Io)?;

    let mut buffered = BufferedStream::new(stream);
    let frame = match tokio::time::timeout(timeout, buffered.read_frame(262_144)).await {
        Ok(Ok(f)) => f,
        Ok(Err(e)) => return Err(PingError::Io(e)),
        Err(_) => return Err(PingError::Timeout),
    };
    let (_length, header_len) = read_var_int(&frame).map_err(|_| PingError::Io(std::io::Error::new(std::io::ErrorKind::InvalidData, "bad frame")))?;
    let (packet_id, consumed) =
        read_var_int(&frame[header_len..]).map_err(|_| PingError::Io(std::io::Error::new(std::io::ErrorKind::InvalidData, "bad packet id")))?;
    if packet_id != 0x00 {
        return Err(PingError::UnexpectedPacketId(packet_id));
    }
    let (json, _) =
        read_string(&frame[header_len + consumed..], 262_144).map_err(|_| PingError::Io(std::io::Error::new(std::io::ErrorKind::InvalidData, "bad string")))?;
    Ok(PingResult { status_json: json, latency_millis: start.elapsed().as_millis() as i64 })
}

#[cfg(test)]
mod tests {
    use super::*;
    use tokio::io::AsyncReadExt;
    use tokio::net::TcpListener;

    #[tokio::test]
    async fn ping_backend_live_success() {
        let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
        let addr = listener.local_addr().unwrap();
        tokio::spawn(async move {
            let (mut sock, _) = listener.accept().await.unwrap();
            let mut buf = vec![0u8; 4096];
            let n = sock.read(&mut buf).await.unwrap();
            let _ = n; // consume handshake + status request (arrives as one or two reads)
            let mut buf2 = vec![0u8; 4096];
            let _ = tokio::time::timeout(Duration::from_millis(100), sock.read(&mut buf2)).await;
            let json = r#"{"version":{"name":"1.21","protocol":767}}"#;
            let mut payload = Vec::new();
            write_var_int(&mut payload, 0x00);
            let mut full = Vec::new();
            write_var_int(&mut full, json.len() as i32);
            full.extend_from_slice(json.as_bytes());
            payload.extend_from_slice(&full);
            let mut frame = Vec::new();
            write_var_int(&mut frame, payload.len() as i32);
            frame.extend_from_slice(&payload);
            tokio::io::AsyncWriteExt::write_all(&mut sock, &frame).await.unwrap();
        });

        let result = ping_backend_live(addr, 767, "example.com", addr.port(), Duration::from_secs(2), false).await.unwrap();
        assert!(result.status_json.contains("767"));
    }

    #[tokio::test]
    async fn ping_backend_live_connect_failure() {
        // Port 0 after binding-and-dropping is unlikely to be listening; use an address nothing
        // listens on instead for a deterministic connection refusal.
        let addr: SocketAddr = "127.0.0.1:1".parse().unwrap();
        let result = ping_backend_live(addr, 767, "example.com", 1, Duration::from_millis(500), false).await;
        assert!(result.is_err());
    }
}
